osquery/main/windows/main.cpp

```diff
@@ -342,35 +287,28 @@ Status uninstallService() {
 void WINAPI ServiceControlHandler(DWORD control_code) {
   switch (control_code) {
   case SERVICE_CONTROL_STOP:
-  case SERVICE_CONTROL_SHUTDOWN:
+  case SERVICE_CONTROL_SHUTDOWN: {
     if (kServiceStatus.dwCurrentState != SERVICE_RUNNING) {
       break;
     }
 
     // Give the main thread a chance to shutdown gracefully before exiting
     UpdateServiceStatus(0, SERVICE_STOP_PENDING, 0, 3, kServiceShutdownWait);
-    {
-      auto stopEvent = osquery::getStopEvent();
-      if (stopEvent != nullptr) {
-        auto ret = SetEvent(stopEvent);
-        if (ret != TRUE) {
-          SLOG("SetEvent failed (lasterror=" + std::to_string(GetLastError()) +
-               ")");
-        }
-        CloseHandle(stopEvent);
-      }
-      auto thread = OpenThread(SYNCHRONIZE, false, kLegacyThreadId);
-      if (thread != nullptr) {
-        WaitForSingleObjectEx(thread, INFINITE, FALSE);
-        CloseHandle(thread);
-      } else {
-        SLOG("Failed to open handle to main thread of execution with " +
-             std::to_string(GetLastError()));
-      }
+
+    Initializer::requestShutdown();
+    auto thread = OpenThread(SYNCHRONIZE, false, kLegacyThreadId);
+    if (thread != nullptr) {
+      WaitForSingleObjectEx(thread, INFINITE, FALSE);
+      CloseHandle(thread);
+    } else {
+      SLOG("Failed to open handle to main thread of execution with " +
```

Comment:
I recommend:
```
auto le = GetLastError();
SLOG(..., std::to_string(le));
```

Reason for doing things this way is that something in this code might update the last error value in the thread, before the call to GetLastError() executes, and you might receive some other error or 0 (`NOERROR`) and wonder why the heck it could have failed.

Reply:
Sure, I can update these while I am here refactoring!